trl/data_utils.py

```diff
@@ -539,21 +539,29 @@ def _pack_ffd(examples: pa.Table, seq_length: int) -> pa.Table:
 
             # Reconstruct packed values more efficiently
             values_numpy = values.to_numpy()
-            packed_values = []
-            new_offsets = [0]
+            packed_values: list[np.dtype] = []
+            new_offsets: list[int] = [0]
+            sequence_lengths: list[list[int]] = []
 
             for _, seq_indices in bins:
+                seq_lens_in_bin = []
                 for seq_idx in seq_indices:
                     _, start, end = sequences[seq_idx]
                     packed_values.extend(values_numpy[start:end])
+                    seq_lens_in_bin.append(end - start)
+                sequence_lengths.append(seq_lens_in_bin)
                 new_offsets.append(len(packed_values))
 
             dtype = offsets.type.to_pandas_dtype()
             new_offsets = np.array(new_offsets, dtype=dtype)
             packed_values = pa.array(packed_values, type=values.type)
+            sequence_lengths = pa.array(sequence_lengths, type=pa.list_(pa.int32()))
             column = type(column).from_arrays(new_offsets, packed_values)
-        packed_columns.append(column)
-    return pa.Table.from_arrays(packed_columns, names=examples.column_names)
+            packed_columns.append(column)
+            packed_columns.append(sequence_lengths)
+        else:
+            packed_columns.append(column)
```

Comment:
This should work as well?

```suggestion
            packed_columns.append(sequence_lengths)
        packed_columns.append(column)
```

Reply:
actually it seems like `sequence_lengths` can be appended more than once, if you've more than one col that matches `if pyarrow.types.is_list(column.type) or pyarrow.types.is_large_list(column.type):`

Reply:
updated